p2p: Ensure transaction announcements are only queued for fully connected peers

dergoegge opened the pull request:
`TxRelay::m_next_inv_send_time` is initialized to 0, which means that any txids in `TxRelay::m_tx_inventory_to_send` will be announced on the first call to `PeerManagerImpl::SendMessages` for a fully connected peer (i.e. it completed the version handshake).

Prior to #21160, `TxRelay::m_tx_inventory_to_send` was guaranteed to be empty on the first `SendMessages` call, as transaction announcements were only queued for fully connected peers. #21160 replaced a `CConnman::ForEachNode` call with a loop over `PeerManagerImpl::m_peer_map`, in which the txid for a transaction to be relayed is added to `TxRelay::m_tx_inventory_to_send` for all peers. Even for those peers that have not completed the version handshake. Prior to the PR this was not the case as `ForEachNode` has a "fully connected check" before calling a function for each node.

dergoegge committed:
[test] Add p2p_tx_privacy.py

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2005,18 +2005,29 @@ void PeerManagerImpl::SendPings()
-        if (!tx_relay) continue;
+    m_connman.ForEachNode([this, &wtxid, &txid](CNode* node) {
+        PeerRef peer{this->GetPeerRef(node->GetId())};
+        if (!peer) return false;
```
```suggestion
        if (!peer) return;
```

nit: The function return type is `void`

MarcoFalke commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,81 @@
+        # Spy sends version message to enable tx relay but doesn't send a verack
+        txrelay_version_msg = msg_version()
+        txrelay_version_msg.nVersion = P2P_VERSION
+        txrelay_version_msg.strSubVer = "😈😈😈"
```
Most of this code in the test can be removed:


```diff
diff --git a/test/functional/p2p_tx_privacy.py b/test/functional/p2p_tx_privacy.py
index 7944095abc..d7139f5ab7 100755
--- a/test/functional/p2p_tx_privacy.py
+++ b/test/functional/p2p_tx_privacy.py
@@ -17,54 +17,29 @@ from test_framework.p2p import (
     p2p_lock,
 )
 from test_framework.test_framework import BitcoinTestFramework
+from test_framework.wallet import MiniWallet
 
 class P2PTxSpy(P2PInterface):
-    def __init__(self, wtxidrelay=True):
-        super().__init__(wtxidrelay=wtxidrelay)
-
     def on_version(self, message):
         # Avoid sending verack in response to version.
         # When calling add_p2p_connection, wait_for_verack=False must be set (see
         # comment in add_p2p_connection).
-        if message.nVersion >= 70016 and self.wtxidrelay:
-            self.send_message(msg_wtxidrelay())
+        self.send_message(msg_wtxidrelay())
 
 class TxPrivacyTest(BitcoinTestFramework):
     def set_test_params(self):
-        self.setup_clean_chain = True
         self.num_nodes = 1
 
     def run_test(self):
-        self.generate(self.nodes[0], 200)
+        self.wallet = MiniWallet(self.nodes[0])
+        self.wallet.rescan_utxos()
 
         tx_originator = self.nodes[0].add_p2p_connection(P2PInterface())
-        spy = self.nodes[0].add_p2p_connection(P2PTxSpy(), send_version=False, wait_for_verack=False)
-
-        # Spy sends version message to enable tx relay but doesn't send a verack
-        txrelay_version_msg = msg_version()
-        txrelay_version_msg.nVersion = P2P_VERSION
-        txrelay_version_msg.strSubVer = "😈😈😈"
-        txrelay_version_msg.nServices = P2P_SERVICES
-        txrelay_version_msg.relay = 1
-        spy.send_message(txrelay_version_msg)
+        spy = self.nodes[0].add_p2p_connection(P2PTxSpy(), wait_for_verack=False)
         spy.wait_for_verack()
 
         # Create a tx to send
-        prevtx = self.nodes[0].getblock(self.nodes[0].getblockhash(5), 2)['tx'][0]
-        rawtx = self.nodes[0].createrawtransaction(
-            inputs=[{'txid': prevtx['txid'], 'vout': 0}],
-            outputs=[{self.nodes[0].get_deterministic_priv_key().address: 50 - 0.00125}],
-        )
-        sigtx = self.nodes[0].signrawtransactionwithkey(
-            hexstring=rawtx,
-            privkeys=[self.nodes[0].get_deterministic_priv_key().key],
-            prevtxs=[{
-                'txid': prevtx['txid'],
-                'vout': 0,
-                'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
-            }],
-        )['hex']
-        tx = tx_from_hex(sigtx)
+        tx = self.wallet.create_self_transfer()["tx"]
 
         # tx_originator sends a tx
         tx_originator.send_and_ping(msg_tx(tx))

dergoegge replied:
Amazing! Done.

maflcko reviewed (commented):
good catch

jnewbery commented:
Concept ACK.

Did you consider setting `m_next_inv_send_time` to a time point in the future inside `SetTxRelay()`? That seems like a simpler change, and avoids adding a new instance of `ForEachNode()` which we'd like to get rid off eventually.

dergoegge commented:
@jnewbery Really good question! I should have put my reasoning for this in the PR description. 

> Did you consider setting m_next_inv_send_time to a time point in the future inside SetTxRelay()? That seems like a simpler change, and avoids adding a new instance of ForEachNode() which we'd like to get rid off eventually.

This actually was my initial approach, however I ended up going with `ForEachNode` for a couple reasons:

* If we initialize `m_next_inv_send_time` to a time point in the future then we will still announce transactions that were received during the handshake which seems fine from a privacy perspective given the random delay. However,
  *  I could see how that leads to a followup bug in which the spy is able to tell which of the announced transactions were received during the handshake (e.g. see point below). We can eliminate the possibility entirely by going with the `ForEachNode` approach.
  * A node can't know if a peer enables wtxid relay until the handshake completes and it might add txids to `m_tx_inventory_to_send` but announce them as `MSG_WTX` after the handshake. (Didn't test but i think) This again leaks which transactions were received between the spy sending `version` and `wtxidrelay`.
* Storing the txids pre-handshake completion feels a bit weird from a resource consumption perspective. We rely on the 60 second p2p timeout (`-peertimeout`) to clear `m_tx_inventory_to_send` for peers that do not complete the handshake. Again not really an issue now but this doesn't feel robust w.r.t. to future changes to me.

glozow reviewed (commented):
ACK 4dc67b2a82cc71fc8b11247bd0f3028a6cae5494

Agree with the approach to not add to `m_tx_inventory_to_send` until the handshake completes, and that not knowing wtxidrelay yet is also a good reason. Likewise not a fan of `ForEachNode` here but don't see a way around it given only `CNode` knows whether it's fully connected.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2005,18 +2005,27 @@ void PeerManagerImpl::SendPings()
-    for(auto& it : m_peer_map) {
-        Peer& peer = *it.second;
-        auto tx_relay = peer.GetTxRelay();
-        if (!tx_relay) continue;
```
Wouldn't it be better to make this a one-line patch?

`m_next_inv_send_time` will remain `0` for at least as long as the `fSuccessfullyConnected` is false. So an alternative fix would be to check `m_next_inv_send_time==0` in this line.

This should behave the same, because `SendMessages` is run after `fSuccessfullyConnected = true` without processing transactions from different peers in-between.

dergoegge replied:
Hm I kinda prefer the current patch. `m_next_inv_send_time` is not meant to be used as a "successfully connected check" so i prefer not to use it in that way here.

jnewbery replied:
I agree with Marco that it'd be better to use state in the `Peer` object rather than reaching into the net layer. It feels like that'd make it easier to unit test net procesing (using #25515).

Using the special value `m_next_inv_send_time == 0` to mean "we're not queueing up an inv message for this peer so don't save hashes to send" seems ok to me as long as it's commented. Alternatively, it'd be easy enough to add a bool to `TxRelay` to gate the logic for saving hashes.

dergoegge committed:
[net processing] Ensure transaction announcements are only queued for fully connected peers

dergoegge committed:
[net processing] Assume that TxRelay::m_tx_inventory_to_send is empty pre-verack

This commit documents our assumption about
TxRelay::m_tx_inventory_to_send being empty prior to version handshake
completion.

The added Assume acts as testing oracle for our fuzzing tests to
potentially detect if the assumption is violated.

luke-jr reviewed (approved):
ACK 959cc24fcd19270d7d851dfde572c4d5848f2579

mzumsande commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,47 @@
+        tx_originator.send_and_ping(msg_tx(tx))
+
+        with p2p_lock:
+            spy.last_message.pop("inv", None)
```
What is the reason for including this `pop`? If bitcoind ever sent an inv to a peer before receiving a verack, shouldn't this test rather fail than silently remove it?

dergoegge replied:
No reason, removed it. Thanks!

mzumsande reviewed (commented):
Concept ACK

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3428,6 +3433,18 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             }
         }
 
+        if (auto tx_relay = peer->GetTxRelay(); tx_relay) {
```
```suggestion
        if (auto tx_relay = peer->GetTxRelay()) {
```

nit?

dergoegge replied:
Done

maflcko reviewed (approved):
lgtm

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2011,8 +2011,13 @@ void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid
         LOCK(tx_relay->m_tx_inventory_mutex);
+        // Only queue transactions for announcement once the version handshake
+        // is completed. The time of arrival for these transactions is
+        // otherwise at risk of leaking to a spy.
```
I can't fully grasp the exact risk here... How does it help the spy to learn something new, as opposed to listening/analyzing *after* the handshake?

MarcoFalke replied:
There is no delay, so they can poll all announcements in real time by just creating a new incoming connection in a loop

naumenkogs replied:
Yeah, the comment from the following commit also helped :) Feel free to resolve this. This is a great finding.

I can only suggest explaining it here, and not in the following comment, because IMO the justification belongs where the transactions are added.

dergoegge replied:
Improved the comment in this location a little bit.

naumenkogs commented:
Concept ACK.

Looking forward to resolving the comments :)

mzumsande reviewed (commented):
Code Review ACK 1220dfaf7c64f2eebe7c8061634d1f3fd2100a61

If you touch again, could run flake8 or similar over the test (it complains about an unused `p2p_lock` import and some blank lines)

jnewbery commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Perhaps add a module-level docstring to describe what the test is testing, and how it is testing it. (https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md#style-guidelines)

jnewbery commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,44 @@
+        # Spy sends the verack
+        spy.send_and_ping(msg_verack())
+
+        # Spy should not get an inv for the tx
+        assert 'inv' not in spy.last_message
```
Perhaps send a second tx from `tx_originator` to the node _after_ the version handshake with spy is complete, and then asset that the node announces the second tx, but not the first.

jnewbery reviewed (commented):
Looks good. A couple of minor suggestions inline in the functional test.

luke-jr reviewed (approved):
re-ACK 152a42a3344b473c838b94f2ec07b25d6739bd2d

glozow commented on src/net_processing.cpp:
```diff
@@ -295,7 +295,7 @@ struct Peer {
         /** The next time after which we will send an `inv` message containing
          *  transaction announcements to this peer. */
-        std::chrono::microseconds m_next_inv_send_time GUARDED_BY(NetEventsInterface::g_msgproc_mutex){0};
+        std::chrono::microseconds m_next_inv_send_time GUARDED_BY(m_tx_inventory_mutex){0};
```
in 845e3a34c49abcc634b5a10ccdd6b10fb4fcf449, noting to myself that this change is because `m_next_inv_send_time` can now be accessed from the scheduler thread in `ReattemptInitialBroadcast`.

glozow commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,69 @@
+
+  tx_originator ----> node[0] <---- spy
+
+We test that a transaction send by tx_originator is only relayed to spy
```
micro nit in fbcd991f8c9fe6ecdac181b09f04756cab99f47d
```suggestion
We test that a transaction sent by tx_originator is only relayed to spy
```

glozow reviewed (commented):
ACK fbcd991f8c9fe6ecdac181b09f04756cab99f47d

MarcoFalke commented on test/functional/p2p_tx_privacy.py:
```diff
@@ -0,0 +1,69 @@
+
+        # Spy should only get an inv for the second transaction as the first
+        # one was received pre-verack with the spy
+        spy.wait_for_inv([CInv(MSG_WTX, tx2.calc_sha256(True))])
```
The test doesn't test anything. It passes on master as-is. I think you can just revert to the earlier commit where it was working?

jnewbery replied:
I think you could add an `on_inv` method to the spy connection that saves the announced hash, and then assert at the end that you haven't received an annoucement for tx1.

dergoegge replied:
Did pretty much what John suggested. It's now using on_inv to collect all announced invs and in the end it checks that there is only one announcement for tx2.

1440000bytes commented:
Concept ACK

maflcko reviewed (approved):
ACK 8f2dac54096c20afd8fd12c21a2ee5465fea085e 🔝

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8f2dac54096c20afd8fd12c21a2ee5465fea085e 🔝
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUghBAwAyGNtXO8fvuygZ5qpYaJJCN8updyEfpF8Q//EHzHYtA7gyW1HhZKTXASV
7/R41rYKVtSE6XqJ8piivfuYNNTDZmpa1/LX030yZnSJsUGfr28bWPkXxrcUmdNE
CsiwIUbrPAzwfiPciowfipA/7d894Ado8X54ESkADJFZvYMuZ63aUuby/aPa/lCC
4GotuNiYjXJpC8FfeR6VL7yJORVdgMav8jNPzvUYM5DNfGex39Uzh1kACtXrtXTn
zeT3zqrVf/rhr9izKqQKN5utGt3vcK9+8F8zh82NOr4heNG8ZufPWHP8F+D/uQPY
h/qNQhvT47NOODa4KKurLYWSEQXYaYRdOVLXq40lZOPMbU5ByAGwMjNWusemAb4/
Af+Xt7UlK4irsQQOj+D9EClqzllVL0gfqPvDmYYutgr/6co2rgg+MJSg8LHN9t+X
rTe2pwtee+wvLTYZGgKg7uJrcanmrUL87sSFvXutOy0TXbnJ67FR3B5fGCFkXnbJ
wYTAXZpf
=/QGZ
-----END PGP SIGNATURE-----
```


</details>

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3428,6 +3435,20 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            Assume(WITH_LOCK(
+                tx_relay->m_tx_inventory_mutex,
+                return tx_relay->m_tx_inventory_to_send.empty() &&
+                       tx_relay->m_next_inv_send_time == 0s));
```
(not in this PR) it might be nice to set explicitly set `m_next_inv_send_time` to a future time here. It's slightly odd that we go through the transaction sending logic on the first pass through `SendMessages()` even though `m_tx_inventory_to_send` will be empty.

jnewbery reviewed (commented):
utACK 8f2dac54096c20afd8fd12c21a2ee5465fea085e

fanquake commented:
Added this for backport in #26616.

1440000bytes commented:
Love that it was merged in less than 10 days, improves bitcoin and reviewed by more than 6 devs :)
